### DISCUSSION TITLE
focusAtom mixed with async getter?

### DISCUSSION THREAD

**Discussion by mrcwinn** [pmndrs/jotai#1647] [Q&A]
Hi, we're evaluating Jotai for a fairly complex application. A core use case is a best practice to remote-fetching initial state. It may be preferred to fetch all initial state in a single call to avoid repeat network requests, but then that will bias toward more complex objects.

Say I have an atom, the contents of which originates from some remote server.

```
const fancyAtom = atom(
  async () => ({
    products: await (await fetch('/api')).json()
  })
)

const productsAtom = focusAtom(fancyAtom, (optic) => optic.prop('products'))
```

`focusAtom` will fail because `fancyAtom` returns only the promise of a writable atom, not the writable atom itself. Is there a way to use focus atoms when the source relies on an asynchronous operation? Have I overthought this entirely?

**Comment by mrcwinn** [comment]
I think I just caught my silly mistake. By writing my atom in such a way, I effectively made it read-only, as there is no setter. `focusAtom` expects it to be a WritableAtom. D'oh. I'll leave this here in case it's helpful to someone.

**Comment by dai-shi** [comment]
focusAtom works with async atoms.

In the OP example, we can just derive atoms:

```js
const fancyAtom = atom(
  async () => ({
    products: await (await fetch('/api')).json()
  })
)

const productsAtom = atom((get) => get(fancyAtom).products)
```
